File: server/src/main/java/com/adobe/testing/s3mock/dto/ListPartsResult.java
/*
 *  Copyright 2017-2024 Adobe.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.adobe.testing.s3mock.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonRootName;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import java.util.List;

/**
 * List-Parts result with some hard-coded values as this is sufficient for now.
 * <a href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListParts.html">API Reference</a>
 */
@JsonRootName("ListPartsResult")
public record ListPartsResult(
    @JsonProperty("Bucket")
    String bucket,
    @JsonProperty("Key")
    String key,
    @JsonProperty("UploadId")
    String uploadId,
    @JsonProperty("PartNumberMarker")
    String partNumberMarker,
    @JsonProperty("NextPartNumberMarker")
    String nextPartNumberMarker,
    @JsonProperty("IsTruncated")
    boolean truncated,
    @JsonProperty("StorageClass")
    StorageClass storageClass,
    @JsonProperty("Part")
    @JacksonXmlElementWrapper(useWrapping = false)
    List<Part> parts,
    @JsonProperty("Owner")
    Owner owner,
    @JsonProperty("Initiator")
    Owner initiator,
    @JsonProperty("ChecksumAlgorithm")
    ChecksumAlgorithm checksumAlgorithm,
    //workaround for adding xmlns attribute to root element only.
    @JacksonXmlProperty(isAttribute = true, localName = "xmlns")
    String xmlns
) {

  public ListPartsResult {
    partNumberMarker = partNumberMarker == null ? "0" : partNumberMarker;
    nextPartNumberMarker = nextPartNumberMarker == null ? "1" : nextPartNumberMarker;
    storageClass = storageClass == null ? StorageClass.STANDARD : storageClass;
    if (xmlns == null) {
      xmlns = "http://s3.amazonaws.com/doc/2006-03-01/";
    }
  }

  public ListPartsResult(String bucketName,
      String key,
      String uploadId,
      List<Part> parts) {
    this(bucketName, key, uploadId, null, null, false, null, parts,
        null, null, null, null);
  }
}

File: server/src/main/kotlin/com/adobe/testing/s3mock/dto/ListPartsResult.kt
/*
 *  Copyright 2017-2026 Adobe.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.adobe.testing.s3mock.dto

import com.adobe.testing.S3Verified
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonRootName
import tools.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper

/**
 * List-Parts result with some hard-coded values as this is sufficient for now.
 * [API Reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListParts.html)
 */
@S3Verified(year = 2025)
@JsonRootName("ListPartsResult", namespace = S3_NS)
data class ListPartsResult(
  @param:JsonProperty("Bucket", namespace = S3_NS)
  val bucket: String?,
  @param:JsonProperty("ChecksumAlgorithm", namespace = S3_NS)
  val checksumAlgorithm: ChecksumAlgorithm?,
  @param:JsonProperty("ChecksumType", namespace = S3_NS)
  val checksumType: ChecksumType?,
  @param:JsonProperty("Initiator", namespace = S3_NS)
  val initiator: Initiator?,
  @param:JsonProperty("IsTruncated", namespace = S3_NS)
  val isTruncated: Boolean,
  @param:JsonProperty("Key", namespace = S3_NS)
  val key: String?,
  @param:JsonProperty("MaxParts", namespace = S3_NS)
  val maxParts: Int?,
  @param:JsonProperty("NextPartNumberMarker", namespace = S3_NS)
  val nextPartNumberMarker: Int?,
  @param:JsonProperty("Owner", namespace = S3_NS)
  val owner: Owner?,
  @param:JacksonXmlElementWrapper(useWrapping = false)
  @param:JsonProperty("Part", namespace = S3_NS)
  val parts: List<Part>?,
  @param:JsonProperty("PartNumberMarker", namespace = S3_NS)
  val partNumberMarker: Int?,
  @param:JsonProperty("StorageClass", namespace = S3_NS)
  val storageClass: StorageClass?,
  @param:JsonProperty("UploadId", namespace = S3_NS)
  val uploadId: String?,
)
